### ISSUE TITLE
Some segments not being pre-processed if there is a whitespace in the definition

### ISSUE THREAD

**Issue by iMarkus** [piwik/piwik#10973]
Hi,
we are currently facing issues with Piwik 2.16.2

The widgets do not show any contents when the following segment is used:

pageUrl=@site.domain.com%2Flocation%2Fxyz%2Fen-us%2Forg%2Faaa%2Fbbb%20ccc%20ddd%20eee

but it displays data when using:

pageUrl=^site.domain.com%2Flocation%2Fwxyz%2Fen-us%2Forg%2Faaa%2Fbbb

So it looks like %20 is not working correctly for segmentation process. Also exact matches with blanks in the URL are not working. Example:

pageUrl==site.domain.com%2flocation%2fxyz%2fen-us%2forg%2faaa%2fbbb%20ccc%20ddd%20eee%2fpages%2ftools.aspx

we also tried to use + instead of %20 with no difference

pageUrl==site.domain.com%2flocation%2fxyz%2fen-us%2forg%2faaa%2fbbb+ccc+ddd+eee%2fpages%2ftools.aspx

Similar to this issue:
https://github.com/piwik/piwik/issues/4827#issuecomment-265122956

**Comment by StiftungAusNachlass** [comment]
Hi,
we are facing very similar issues with

Piwik Version: 3.0.1
MySQL Version: 5.5.54
PHP-Version: 5.5.9-1ubuntu4.20

It´s not only the %20 - Blanks. At least also the "/"-Slash is affected.

Directly after saving the segment, the widgets show the expected data. But only once. After reload or anything else, the data won't ever appear again. 
But: It appears again, if we "resave" the segment - even without changes. But again - only once. Reloading hides the data forever..

The segments don't show any data, when a "/" is part of the for example "Download-URL" or "Site-URL". If we reduce the url to a string without "/" or blanks, than it works as expected and not only once..

The problem affects all type of users.

Thanks

**Comment by iMarkus** [comment]
Hi,
The issue is with how Piwik url-decodes and encodes segments' definitions. There are several issues on Github, for example this one: https://github.com/piwik/piwik/issues/10126. Unfortunately, it is another complex Piwik Core bug :(

A workaround, in this case, would be to urlencode what is already urlencoded. So, for example, all '%20' change into '%2520', but characters that are not already urlencoded shouldn't be urlencoded (for example '/' should be still '/').

Hope this will be fixed for version 2.x and 3.x :)

Best regards,
iMarkus

**Comment by mattab** [comment]
it may be worth trying again your issue after https://github.com/piwik/piwik/issues/11321 will be fixed.

as you mention you were client of Piwik pro, they were not be able to help you there and propose a solution for this?

**Comment by mattab** [comment]
Please try our next beta tomorrow and check if it's fixed. We have fixed another segment encoding issue which may also help your issue here

**Comment by iMarkus** [comment]
Hi mattab,

we received the workaround from piwik pro. Looking forward to piwik-3.0.2-b5.

Best regards,
Markus

**Comment by mattab** [comment]
The release has been made so please try it and reopen if you still experience the issue

**Comment by mattab** [comment]
The release has been made so please try it and reopen if you still experience the issue

**Comment by iMarkus** [comment]
Hi mattab,

thanks so far. We will take a backup from our current database and update it to 3.0.2-b5. But it will take some time to prepare this.

I saw you marked it as "cannot reproduce yet". Therefore it would be nice to have an example in Piwik demo to reproduce and test it.

Best regards,
Markus

**Comment by iMarkus** [comment]
Hi again,

We now did the following:

- made a copy of our database
- setup a development environment just for this issue using the same version 2.16.2
- converted to Piwik newest Version 3.0.3
- deleted all archives
- started archiving process

Unfortunately the problem is still existing :(

Best regards,
Markus

**Comment by mattab** [comment]
@iMarkus any chance you could give us access to your PIwik? If so, please contact us via: https://piwik.org/contact/ and send the piwik url + login + password +link to this issue. Thanks

**Comment by mattab** [comment]
please re-open if issue still active

**Comment by dshiryaev-plesk** [comment]
Hello,
we faced this issue with standalone Matomo 3.5.1. We tried to create the segment with the following conditions without luck:
- page URL is 'https://corp.com/path/with+some+spaces'
- page URL is 'https://corp.com/path/with some spaces'
- page URL is 'https://corp.com/path/with%20some%20spaces'
- page URL contains 'https://corp.com/path/with some spaces'
- page URL contains '/corp.com/path/with some spaces'

It works with the following condition but it is not what we want:
- page URL starts with 'https://corp.com/path/with'
